rfc-154-integrating-an-ots-cms-with-govuk-for-custom-pages.md

```diff
@@ -0,0 +1,60 @@
+
+It is common that headless CMS tools make use of [webhooks](https://en.wikipedia.org/wiki/Webhook) as a means to notify subscribing software to events that occur in the CMS. We propose the introduction of a piece of software that will listen for a CMS's webhooks and use this information as an initiator of a process to update the content.
+
+This software will identify whether a webhook event signifies a live or draft event and update the Publishing API accordingly.
+
+For integration with Contentful, we will deploy a new GOV.UK application: Contentful Listener API. This was [developed](https://github.com/alphagov/contentful-listener-api) whilst spiking approaches to this problem. There is documentation written to the describe the [workflow](https://github.com/alphagov/contentful-listener-api/blob/main/docs/how-this-application-works.md) of this tool and [integration limitations](https://github.com/alphagov/contentful-listener-api/blob/main/docs/integration-limitations.md). This application will be owned by the Publishing Platform team.
+
+### Changes to existing software
+
+It is expected that in order to flexibly model a GOV.UK page in an off-the-shelf headless CMS there will need to be multiple content entities involved (ie. to model the Cost of living page, there may be multiple accordion entities associated with a Cost of living page entity). In order to track whether these are associated with content a new field will be added to the Publishing API: `cms_entity_ids` this can be used to store the unique IDs of items in the off-the-shelf headless CMS. When a webhook event is received this ID will be queried to determine if any GOV.UK content is affected.
+
+The [`special_route`](https://github.com/alphagov/govuk-content-schemas/blob/main/formats/special_route.jsonnet) schema in [govuk-content-schemas](https://github.com/alphagov/govuk-content-schemas) will be updated to accept a details field. This details field will not have restrictions on how the data is structured, this will allow changes in the modelling of content without needing a schema to be updated (this was an [approach explored](https://github.com/alphagov/govuk-content-schemas/blob/170a941b42bbc53d529b4c50698ad5ba50776df5/formats/coronavirus_landing_page.jsonnet#L4-L8) for the Coronavirus landing page). It is theorised that having a schema for a custom page is more of a hindrance than benefit given the data is only used in a single place.
+
+### Further considerations
+
+- If the CMS was offline and we needed to make edits to content this would be done by a short term hardcoding in the frontend application to set content.
+- It is unlikely, due to complexity reasons, that we would integrate [Asset Manager](https://github.com/alphagov/asset-manager) with the CMS tooling and instead rely on the CMS to host assets such as images. If the CMS was offline these would be unavailable, but as the page would still be available this would not be a critical issue to resolve.
```

Comment:
I feel a bit unsure about this, although obviously I haven't looked into the complexity. Feels a bit of a shame that we're using the common backend for the content, but not for the assets. I guess I'm thinking of 
* Adding complexity to the stack
* Having to migrate if we change tack
Obviously not a big deal if it's only ends up being one page, but thinking of this being used more widely.

Reply:
As the Publishing API doesn’t have a relationship with asset manager the only thing we’d do if we had an asset manager file is replace the CMS URL in the content item with the asset manager one.

We’d then lose the ability to utilise any extra asset APIs, for example with Contentful we can use their [image API](https://www.contentful.com/developers/docs/references/images-api/) for dynamic resizing.

The one thing that I feel is a loss with using the CMS assets is the hostname (I like vanity URLs :laughing:) as with Contenful they’d be on cfassets.com. However I feel like, if we had to resolve that, I’d prefer to proxy to the CMS’ asset server off one of our hostnames.

Reply:
Thanks Ollie, some really interesting points.

> 1. Avoid breaking links to assets

This situation seems a concern if non-image files are uploaded for downloading, but doesn't seem to be too much of an issue for an image file. In the discussed use cases for this images seemed the only likely use case. I'm not too concerned if they break, as they feel like sub-resources.

> 2. Incomplete archives on The National Archives

This is a good point. It'd be worth finding out their policies. I wasn't aware this would be a restriction

> 3. Lack of control

Do agree it seems a bit annoying to lose control. This seems like something we'd only have limited control of with a proxy too.

---

I did ponder that setting up a proxy might be more delicate than I considered. We'd need to only allow proxying particular URLs of a CMS otherwise it could look like GOV.UK was hosting assets for other clients. I guess taking control of any rogue URLs would be a bit of pain as it'd be done through nginx config - but I imagine it's so rare it's not too big a deal.

---

Hmm tricky. I'm imagining the easiest resolution if there's not a clear route would be to defer a decision on what to do with assets until we learn more about if the experiment is successful, as it's not a clear need yet and thus likely won't get a decent amount of attention.

I'd think that considering Contentful, a proxying approach would be a good option for keeping things like image resizing, which I imagine could be frustrating to not have. For the issues you've outlined though an Asset Manager integration would be better. Though I imagine doing it would be a bit of a headache in trying to determine whether an asset needs updating or not.

Reply:
This does feel like the most contentious part of the proposal. A few points:

1) Agreed that proxying to the CMS asset URL opens up a bit of a security risk in terms of bad actors hosting offensive content and being able to craft GOV.UK URLs that link to said content.

2) Linking directly to `cfassets.com` also not ideal, I'm thinking of people who may have script blockers where they've allow-listed GOV.UK but not third party providers. A small point, but worth considering.

3) The most important point IMO is the migration strategy - it seems highly likely we'd eventually move away from whichever CMS is picked for this experiment, and then be left with a migration job of having to copy all the images to GOV.UK's asset manager and rewrite all the affected content in Publishing API to refer to the new image names.
We also may not actually _want_ to utilise the dynamic image sizing functionality in the Contentful Images API, as it would likely increase the complexity of the integration with GOV.UK's stack.

One thought that came to mind was building some sort of simple asset uploader CMS, where publishers can upload assets directly to Asset Manager, and then copy the resulting asset URL into the headless CMS. A bit more clunky, perhaps, but less of a headache when it comes to migration.
Would need to think about how to handle alt text - the asset uploader CMS feels like a reasonable fit for that, but it would be another thing to have to copy and paste, so probably best handled within the headless CMS.

Reply:
Thanks for your points, I'm thinking the simplest thing is perhaps to defer asset handling for now and to have it as a limitation - perhaps just not include assets in JSON payloads so they can't be snuck in.

> Linking directly to cfassets.com also not ideal, I'm thinking of people who may have script blockers where they've allow-listed GOV.UK but not third party providers. A small point, but worth considering.

Would we expect that for images / assets? Do we have evidence of these existing?

> The most important point IMO is the migration strategy - it seems highly likely we'd eventually move away from whichever CMS is picked for this experiment, and then be left with a migration job of having to copy all the images to GOV.UK's asset manager and rewrite all the affected content in Publishing API to refer to the new image names.
We also may not actually want to utilise the dynamic image sizing functionality in the Contentful Images API, as it would likely increase the complexity of the integration with GOV.UK's stack.

This doesn't concern me too much as I'm expecting the initial footprint to be pretty low, so relatively solvable should we make the decision to drop/change.

> One thought that came to mind was building some sort of simple asset uploader CMS, where publishers can upload assets directly to Asset Manager, and then copy the resulting asset URL into the headless CMS. A bit more clunky, perhaps, but less of a headache when it comes to migration.
Would need to think about how to handle alt text - the asset uploader CMS feels like a reasonable fit for that, but it would be another thing to have to copy and paste, so probably best handled within the headless CMS.

This is a good idea, but I imagine it'd be less work to set-up a system that syncs a CMS assets with Asset Manager so would probably go that way. I imagine it'd be less confusing too than having to forgo a CMS assets features.

Manually uploading assets does seem a reasonable workaround though in the early days if this feature is needed.